src/markdown.py

```diff
@@ -37,7 +37,19 @@ def convert_markdown_to_html(markdown: str) -> str:
     markdown = re.sub(r"## (.*?)\n", r"<h2>\1</h2>\n", markdown)
     markdown = re.sub(r"# (.*?)\n", r"<h1>\1</h1>\n", markdown)
 
+    # Convert all head space to &nbsp;
+    markdown = convert_leading_spaces_to_html(markdown)
+
     # Convert newlines to <br> tags
     markdown = re.sub(r"\n", r"<br>", markdown)
 
     return markdown
+
+
+# Convert leading spaces to &nbsp;
+def convert_leading_spaces_to_html(markdown):
+    def replace_spaces(match):
+        spaces = match.group(1)
+        return '&nbsp;' * len(spaces) + match.group(2)
+
+    return re.sub(r"^( +)(.*)", replace_spaces, markdown, flags=re.MULTILINE)
```

Comment:
Any particular reason why we match the entire string with `^( +)(.*)` versus just leading matches with `^ +`? Is this to avoid the case of converting lines which contain only spaces?

Reply:
I added some replacement examples, could you please take a look again?

Indeed, both can achieve the same functionality.

However, after matching the leading whitespace characters, I don't know how to replace them with an equivalent amount of nbsp. I found the code provided by GPT quite suitable for my scenario, and its functionality meets my needs, so I used it like that.

I'm not particularly familiar with Python; I only know the basics. 😆

Reply:
Taking another look! I think the `* len(spaces)` is very readable tbh, just not sure we need the second `match.group(2)`. Testing right now

Reply:
Let me try it. 😆

Reply:
I have made the adjustments. Please take a look when you have time.

Reply:
Just tested it, looks great. Merging!